crates/key-value/src/util.rs

```diff
@@ -31,3 +39,168 @@ impl StoreManager for DelegatingStoreManager {
+/// cross-cache consistency subject to scheduling and/or networking delays, a given tuple is never refreshed from
+/// the backing store once added to a cache since this implementation is intended for use only by short-lived guest
+/// instances.
+///
+/// Note that, because writes are asynchronous and return immediately, durability is _not_ guaranteed.  I/O errors
+/// may occur asyncronously after the write operation has returned control to the guest, which may result in the
+/// write being lost without the guest knowing.  In the future, a separate `write-durable` function could be added
+/// to key-value.wit to provide provide either synchronous or asynchronous feedback on durability for guests which
+/// need it.
+pub struct CachingStoreManager<T> {
+    inner: T,
+}
+
+impl<T> CachingStoreManager<T> {
+    pub fn new(inner: T) -> Self {
+        Self { inner }
+    }
+}
+
+#[async_trait]
+impl<T: StoreManager> StoreManager for CachingStoreManager<T> {
+    async fn get(&self, name: &str) -> Result<Arc<dyn Store>, Error> {
+        Ok(Arc::new(CachingStore {
+            inner: self.inner.get(name).await?,
+            state: AsyncMutex::new(CachingStoreState::default()),
+        }))
+    }
+}
+
+#[derive(Default)]
+struct CachingStoreState {
+    cache: HashMap<String, Option<Vec<u8>>>,
+    previous_task: Option<JoinHandle<Result<(), Error>>>,
+}
+
+impl CachingStoreState {
+    /// Wrap the specified task in an outer task which waits for `self.previous_task` before proceeding, and spawn
+    /// the result.  This ensures that write order is preserved.
+    fn spawn(&mut self, task: impl Future<Output = Result<(), Error>> + Send + 'static) {
+        let previous_task = self.previous_task.take();
+        self.previous_task = Some(task::spawn(async move {
+            if let Some(previous_task) = previous_task {
+                previous_task
+                    .await
+                    .map_err(|e| Error::Io(format!("{e:?}")))??
+            }
+
+            task.await
+        }))
+    }
+}
+
+struct CachingStore {
+    inner: Arc<dyn Store>,
+    state: AsyncMutex<CachingStoreState>,
+}
+
+#[async_trait]
+impl Store for CachingStore {
+    async fn get(&self, key: &str) -> Result<Vec<u8>, Error> {
+        // Retrieve the specified value from the cache, lazily populating the cache as necessary.
+
+        match self.state.lock().await.cache.entry(key.to_owned()) {
+            Entry::Vacant(e) => {
+                let value = match self.inner.get(key).await {
+                    Ok(value) => Some(value),
+                    Err(Error::NoSuchKey) => None,
+                    e => return e,
+                };
+
+                e.insert(value.clone());
+
+                value
+            }
+            Entry::Occupied(e) => e.get().clone(),
+        }
+        .ok_or(Error::NoSuchKey)
+    }
+
+    async fn set(&self, key: &str, value: &[u8]) -> Result<(), Error> {
```

Comment:
Somewhat an aside, but I wonder if the trait could be changed here to allow `key` and `value` to be passed in already-owned rather than needing to `to_owned` them in here, like a `Cow<str>`. On the other hand, the compiler might be able to optimize away the clone anyway, so perhaps not worth the type noise.

Reply:
Yeah, there's a lot of allocation and copying in the current implementation, due to needing to bundle things up into a `'static` `Future`.  I agree `Cow` could be a good option.